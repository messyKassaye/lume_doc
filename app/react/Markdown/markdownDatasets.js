import HtmlParser from 'htmlparser2/lib/Parser';
import queryString from 'query-string';
import rison from 'rison-node';
import Big from 'big.js';

import searchApi from 'app/Search/SearchAPI';
import api from 'app/utils/api';
import entitiesApi from 'app/Entities/EntitiesAPI';

const conformUrl = ({ url = '', geolocation = false }) => {
  const { q } = queryString.parse(url.substring(url.indexOf('?')));

  if (!q) {
    const defaultValue = geolocation
      ? { allAggregations: true, limit: 0, geolocation: true }
      : { allAggregations: true, limit: 0 };

    return defaultValue;
  }

  const params = rison.decode(q);
  params.limit = 0;

  if (geolocation) {
    params.geolocation = true;
  }

  return params;
};

const conformValues = attribs => (attribs.entity ? attribs : conformUrl(attribs));

const parseDatasets = markdown => {
  const result = {};
  const parser = new HtmlParser(
    {
      onopentag(name, attribs) {
        if (name === 'dataset') {
          result[attribs.name || 'default'] = conformValues(attribs);
        }
        if (name === 'query') {
          result[attribs.name || 'default'] = { url: attribs.url, query: true };
        }
      },
    },
    { decodeEntities: true }
  );

  parser.parseComplete(markdown);
  return result;
};

const requestDatasets = (datasets, requestParams) =>
  Promise.all(
    Object.keys(datasets).map(name => {
      if (datasets[name].query) {
        return api.get(datasets[name].url, requestParams).then(data => ({ data: data.json, name }));
      }
      const apiAction = datasets[name].entity ? entitiesApi.get : searchApi.search;
      const params = datasets[name].entity ? { sharedId: datasets[name].entity } : datasets[name];
      const postAction = datasets[name].entity ? d => d[0] : d => d;
      return apiAction(requestParams.set(params))
        .then(postAction)
        .then(data => ({ data, name }));
    })
  );

const conformDatasets = sets => sets.reduce((memo, set) => ({ ...memo, [set.name]: set.data }), {});

const getAggregations = (state, { property, dataset = 'default' }) => {
  const data = state.page.datasets.get(dataset);
  return !data ? undefined : data.getIn(['aggregations', 'all', property, 'buckets']);
};

const addValues = (aggregations, values) => {
  let result = new Big(0);
  values.forEach(key => {
    const value = aggregations.find(bucket => bucket.get('key') === key);
    const filteredValue = value ? value.getIn(['filtered', 'doc_count']) : 0;
    result = result.plus(filteredValue || 0);
  });
  return Number(result);
};

export default {
  async fetch(markdown, requestParams) {
    const datasets = parseDatasets(markdown);
    return requestDatasets(datasets, requestParams).then(conformDatasets);
  },

  getRows(state, { dataset = 'default' }) {
    const data = state.page.datasets.get(dataset);
    if (!data) {
      return undefined;
    }
    return data.get('rows');
  },

  getAggregations,

  getAggregation(state, { uniqueValues, property, value, dataset = 'default' }) {
    const aggregations = getAggregations(state, { property, dataset });
    if (!aggregations) {
      return undefined;
    }

    if (uniqueValues) {
      return aggregations.filter(a => a.getIn(['filtered', 'doc_count']) !== 0).size;
    }

    const values = value ? value.split(',') : [''];
    return addValues(aggregations, values);
  },

  getMetadataValue(state, { property, dataset = 'default' }) {
    const data = state.page.datasets.get(dataset);
    const propertyExists = data && data.hasIn(['metadata', property]);
    const mos = propertyExists ? data.getIn(['metadata', property]).toJS() : [];
    return mos && mos.length && mos[0].value ? Number(mos[0].value) : undefined;
  },
};
