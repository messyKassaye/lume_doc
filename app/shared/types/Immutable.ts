import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
