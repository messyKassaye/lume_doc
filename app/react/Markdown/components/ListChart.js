import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import rison from 'rison-node';
import queryString from 'query-string';

import Loader from 'app/components/Elements/Loader';
import { arrayUtils } from 'app/Charts';
import MarkdownLink from './MarkdownLink';
import markdownDatasets from '../markdownDatasets';

export const ListChartComponent = props => {
  const { excludeZero, property, data, classname, context, colors } = props;
  const sliceColors = colors.split(',');

  let output = <Loader />;

  if (data) {
    const formattedData = arrayUtils.sortValues(
      arrayUtils.formatDataForChart(data, property, {
        excludeZero: Boolean(excludeZero),
        context,
      })
    );
    let query = { filters: {} };

    if (props.baseUrl) {
      const { q } = queryString.parse(props.baseUrl.substring(props.baseUrl.indexOf('?')));
      query = rison.decode(q);
      query.filters = query.filters || {};
    }

    output = (
      <ul>
        {formattedData.map((item, index) => {
          const Content = (
            <div>
              <div
                className="list-bullet"
                style={{ backgroundColor: sliceColors[index % sliceColors.length] }}
              >
                <span>{item.results}</span>
              </div>
              <span className="list-label">{item.label}</span>
            </div>
          );

          query.filters[property] = { values: [item.id] };

          return (
            <li key={item.id}>
              {props.baseUrl && (
                <MarkdownLink url={`/library/?q=${rison.encode(query)}`} classname="list-link">
                  {Content}
                </MarkdownLink>
              )}
              {!props.baseUrl && Content}
            </li>
          );
        })}
      </ul>
    );
  }

  return <div className={`ListChart ${classname}`}>{output}</div>;
};

ListChartComponent.defaultProps = {
  context: 'System',
  excludeZero: false,
  classname: '',
  colors: '#ffcc00,#ffd633,#ffe066,#ffeb99,#fff5cc',
  data: null,
  baseUrl: null,
};

ListChartComponent.propTypes = {
  property: PropTypes.string.isRequired,
  context: PropTypes.string,
  classname: PropTypes.string,
  colors: PropTypes.string,
  data: PropTypes.instanceOf(Immutable.List),
  baseUrl: PropTypes.string,
  excludeZero: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
};

export const mapStateToProps = (state, props) => ({
  data: markdownDatasets.getAggregations(state, props),
  thesauris: state.thesauris,
});

export default connect(mapStateToProps)(ListChartComponent);
