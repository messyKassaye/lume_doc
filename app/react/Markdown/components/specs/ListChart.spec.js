import React from 'react';

import { shallow } from 'enzyme';
import Immutable from 'immutable';

import { mapStateToProps, ListChartComponent } from '../ListChart.js';
import markdownDatasets from '../../markdownDatasets';

describe('ListChart Markdown component', () => {
  const state = {
    thesauris: Immutable.fromJS([
      {
        _id: 'tContext',
        values: [
          { id: 'id1', label: 'label1' },
          { id: 'id2', label: 'label2' },
          { id: 'id3', label: 'label3' },
        ],
      },
    ]),
  };

  it('should render a placeholder when data is undefined', () => {
    let undefinedValue;
    spyOn(markdownDatasets, 'getAggregations').and.returnValue(undefinedValue);
    const props = mapStateToProps(state, { prop2: 'propValue' });
    const component = shallow(<ListChartComponent {...props} property="prop2" />);

    expect(markdownDatasets.getAggregations).toHaveBeenCalledWith(state, { prop2: 'propValue' });
    expect(component).toMatchSnapshot();
  });

  it('should render the data passed by mapStateToProps', () => {
    spyOn(markdownDatasets, 'getAggregations').and.returnValue(
      Immutable.fromJS([
        { key: 'id1', label: 'label1', filtered: { doc_count: 25 } },
        { key: 'id2', label: 'label2', filtered: { doc_count: 33 } },
        { key: 'missing', label: 'No value', filtered: { doc_count: 45 } },
        { key: 'id3', label: 'label3', filtered: { doc_count: 13 } },
      ])
    );

    const props = mapStateToProps(state, { prop1: 'propValue' });
    const component = shallow(
      <ListChartComponent {...props} property="prop1" classname="custom-class" context="tContext" />
    );

    expect(markdownDatasets.getAggregations).toHaveBeenCalledWith(state, { prop1: 'propValue' });
    expect(component).toMatchSnapshot();
  });
  describe('when excludeZero', () => {
    it('should render without zero values', () => {
      spyOn(markdownDatasets, 'getAggregations').and.returnValue(
        Immutable.fromJS([
          { key: 'id1', label: 'label1', filtered: { doc_count: 25 } },
          { key: 'id2', label: 'label2', filtered: { doc_count: 33 } },
          { key: 'missing', label: 'No value', filtered: { doc_count: 45 } },
          { key: 'id3', label: 'label3', filtered: { doc_count: 0 } },
        ])
      );

      const props = mapStateToProps(state, { prop1: 'propValue' });
      props.excludeZero = 'true';
      const component = shallow(
        <ListChartComponent
          {...props}
          property="prop1"
          classname="custom-class"
          context="tContext"
        />
      );

      expect(markdownDatasets.getAggregations).toHaveBeenCalledWith(state, { prop1: 'propValue' });
      expect(component).toMatchSnapshot();
    });
  });
});
