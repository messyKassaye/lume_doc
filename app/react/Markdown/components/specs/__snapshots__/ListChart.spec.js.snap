// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListChart Markdown component should render a placeholder when data is undefined 1`] = `
<div
  className="ListChart "
>
  <Loader />
</div>
`;

exports[`ListChart Markdown component should render the data passed by mapStateToProps 1`] = `
<div
  className="ListChart custom-class"
>
  <ul>
    <li
      key="id2"
    >
      <div>
        <div
          className="list-bullet"
          style={
            Object {
              "backgroundColor": "#ffcc00",
            }
          }
        >
          <span>
            33
          </span>
        </div>
        <span
          className="list-label"
        >
          label2
        </span>
      </div>
    </li>
    <li
      key="id1"
    >
      <div>
        <div
          className="list-bullet"
          style={
            Object {
              "backgroundColor": "#ffd633",
            }
          }
        >
          <span>
            25
          </span>
        </div>
        <span
          className="list-label"
        >
          label1
        </span>
      </div>
    </li>
    <li
      key="id3"
    >
      <div>
        <div
          className="list-bullet"
          style={
            Object {
              "backgroundColor": "#ffe066",
            }
          }
        >
          <span>
            13
          </span>
        </div>
        <span
          className="list-label"
        >
          label3
        </span>
      </div>
    </li>
  </ul>
</div>
`;

exports[`ListChart Markdown component when excludeZero should render without zero values 1`] = `
<div
  className="ListChart custom-class"
>
  <ul>
    <li
      key="id2"
    >
      <div>
        <div
          className="list-bullet"
          style={
            Object {
              "backgroundColor": "#ffcc00",
            }
          }
        >
          <span>
            33
          </span>
        </div>
        <span
          className="list-label"
        >
          label2
        </span>
      </div>
    </li>
    <li
      key="id1"
    >
      <div>
        <div
          className="list-bullet"
          style={
            Object {
              "backgroundColor": "#ffd633",
            }
          }
        >
          <span>
            25
          </span>
        </div>
        <span
          className="list-label"
        >
          label1
        </span>
      </div>
    </li>
  </ul>
</div>
`;
